apiserver/facades/client/machinemanager/upgrade_series.go

```diff
@@ -0,0 +1,251 @@
+			// gives a typed error to the client, which can then decode ths
+			// optional information later on.
+			results[i].Error = &apiservererrors.UpgradeSeriesValidationError{
+				Cause:  errors.Errorf("upgrade series lock found for %q; series upgrade is in the %q state", machine.Id(), status),
+				Status: status.String(),
+			}
+			continue
+		}
+
+		if err := a.validator.ValidateSeries(entity.Series, machine.Series(), entity.Tag); err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+
+		applications, err := a.state.ApplicationsFromMachine(machine)
+		if err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+
+		if err := a.validator.ValidateApplications(applications, entity.Series, entity.Force); err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+
+		units, err := machine.Units()
+		if err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+
+		unitNames, err := verifyUnits(units)
+		if err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+
+		results[i].UnitNames = unitNames
+	}
+
+	return results, nil
+}
+
+func verifyUnits(units []Unit) ([]string, error) {
+	unitNames := make([]string, len(units))
+	for i, u := range units {
+		agentStatus, err := u.AgentStatus()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if agentStatus.Status != status.Idle {
+			return nil, errors.Errorf("unit %s is not ready to start a series upgrade; its agent status is: %q %s",
+				u.Name(), agentStatus.Status, agentStatus.Message)
+		}
+		unitStatus, err := u.Status()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if unitStatus.Status == status.Error {
+			return nil, errors.Errorf("unit %s is not ready to start a series upgrade; its status is: \"error\" %s",
+				u.Name(), unitStatus.Message)
+		}
+
+		unitNames[i] = u.UnitTag().Id()
+	}
+	return unitNames, nil
+}
+
+type upgradeSeriesState struct {
+	state Backend
+}
+
+func (s upgradeSeriesState) MachineFromTag(tag string) (Machine, error) {
+	machineTag, err := names.ParseMachineTag(tag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	machine, err := s.state.Machine(machineTag.Id())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return machine, nil
+}
+
+func (s upgradeSeriesState) ApplicationsFromMachine(machine Machine) ([]Application, error) {
+	names, err := machine.ApplicationNames()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Ensure that we have unique names for this application request.
+	names = set.NewStrings(names...).Values()
+
+	results := make([]Application, len(names))
+	for i, name := range names {
+		app, err := s.state.Application(name)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		results[i] = app
+	}
+	return results, nil
+}
+
+type upgradeSeriesValidator struct{}
+
+func (s upgradeSeriesValidator) ValidateSeries(requested, machine, tag string) error {
+	return validateSeries(requested, machine, tag)
```

Comment:
`validateSeries` is only called by this method and is in _machinemanager.go_.
I suppose you'll reorganise this when you change it to call out to CharmHub...

Reply:
Indeed I have.